app/xlsfonts/man/xlsfonts.man
.\" -*- coding: us-ascii -*-
.TH XLSFONTS __appmansuffix__ 2022-09-10 __xorgversion__
.SH NAME
xlsfonts \- list fonts available to X server
.SH SYNOPSIS
.ad l
.nf
\fBxlsfonts\fR [\fB\-display\fR \fIhost:dpy\fR] [\fB\-l\fR] [\fB\-ll\fR] [\fB\-lll\fR] [\fB\-m\fR] [\fB\-C\fR] [\fB\-1\fR]
    [\fB\-w\fR \fIwidth\fR] [\fB\-n\fR \fIcolumns\fR] [\fB\-u\fR] [\fB\-o\fR] [\fB\-fn\fR \fIpattern\fR] [\fB\-v\fR]
.fi
.SH DESCRIPTION
.B xlsfonts
lists the fonts available on an X server that match the given pattern.
The wildcard character "*" may be used to match any sequence of characters
(including none), and "?" to match any single character.
If no pattern is given, "*" is assumed.
.PP
The "*" and "?" characters must be quoted to prevent them
from being expanded by the shell.
.SH OPTIONS
.TP
\fB\-display \fIhost:dpy\fB\fR
This option specifies the X server to contact.
.TP
\fB\-d \fIhost:dpy\fB\fR
Same as \fB\-display \fIhost:dpy\fB\fR.
.TP
\fB\-l\fR
Lists some attributes of the font on one line in addition to its name.
.TP
\fB\-ll\fR
Lists font properties in addition to \fB\-l\fR output.
.TP
\fB\-lll\fR
Lists character metrics in addition to \fB\-ll\fR output.
.TP
\fB\-m\fR
This option indicates that long listings should also print
the minimum and maximum bounds of each font.
.TP
\fB\-C\fR
This option indicates that listings should use multiple columns.
This is the same as \fB\-n 0\fR.
.TP
\fB\-1\fR
This option indicates that listings should use a single column.
This is the same as \fB\-n 1\fR.
.TP
\fB\-w \fIwidth\fR
This option specifies the width in characters that should be
used in figuring out how many columns to print.
The default is 79.
.TP
\fB\-n \fIcolumns\fR
This option specifies the number of columns to use in displaying the output.
By default, it will attempt to fit as many columns of font names as possible
into the number of characters specified by \fB\-w\fR \fIwidth\fR.
.TP
\fB\-u\fR
This option indicates that the output should be left unsorted.
.TP
\fB\-o\fR
This option indicates that \fBxlsfonts\fR should do an \fBOpenFont\fR
(and \fBQueryFont\fR, if appropriate) rather than a \fBListFonts\fR.
This is useful if \fBListFonts\fR or \fBListFontsWithInfo\fR fail to
list a known font (as is the case with some scaled font systems).
.TP
\fB\-fn \fIpattern\fR
This option specifies the font name pattern to match.
.TP
.B -v
This option indicates that xlsfonts should print out the program version
and exit.
.SH "SEE ALSO"
.BR X (__miscmansuffix__),
.BR Xserver (__appmansuffix__),
.BR xset (__appmansuffix__),
.BR xfd (__appmansuffix__),
.BR xfontsel (__appmansuffix__),
.I X Logical Font Description Conventions
.SH ENVIRONMENT
.TP
.B DISPLAY
The default host and display to use if neither \fB\-d\fR or \fB\-display\fR
are specified.
.SH BUGS
Doing ``xlsfonts -l'' can tie up your X server for a very long time.
This is really a bug with single-threaded non-preemptable servers,
not with this program.
.SH AUTHOR
Mark Lillibridge
.PP
Jim Fulton
.PP
Phil Karlton
.PP
Roland Mainz


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:

app/xlsfonts/xlsfonts.c
/*

Copyright 1989, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.

 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <X11/Xfuncproto.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xos.h>
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include "dsimple.h"

#ifdef HAVE_BSD_STDLIB_H
#include <bsd/stdlib.h>
#endif

#ifndef HAVE_REALLOCARRAY
#define reallocarray(old, num, size) realloc(old, (num) * (size))
#endif

#define N_START INT_MAX         /* Maximum # of fonts to start with (should
                                 * always be be > 10000 as modern OSes like
                                 * Solaris 8 already have more than 9000 XLFD
                                 * fonts available) */

#define L_SHORT    0
#define L_MEDIUM   1
#define L_LONG     2
#define L_VERYLONG 3

static int  max_output_line_width     = 79;
static int  output_line_padding       = 3;
static int  columns                   = 0;

static Bool sort_output               = True;
static Bool open_instead_of_list      = False;
static int  long_list                 = L_SHORT;
static int  nnames                    = N_START;
static int  font_cnt                  = 0;
static int  min_max;

typedef struct {
    const char *name;
    XFontStruct *info;
} FontList;

static FontList *font_list = NULL;

/* Local prototypes */
static void get_list(const char *pattern);
static int  compare(const void *arg1, const void *arg2);
static void show_fonts(void);
static void copy_number(char **pp1, char **pp2, int n1, int n2);
static int  IgnoreError(Display *disp, XErrorEvent *event);
static void PrintProperty(XFontProp *prop);
static void ComputeFontType(XFontStruct *fs);
static void print_character_metrics(register XFontStruct *info);
static void do_query_font(Display *dpy, const char *name);

void
usage(const char *errmsg, int exitstatus)
{
    if (errmsg != NULL)
        fprintf(stderr, "%s: %s\n\n", program_name, errmsg);

    fprintf(stderr, "usage:  %s [-options] [-fn pattern]\n%s", program_name,
            "where options include:\n"
            "    -l[l[l]]                 give long info about each font\n"
            "    -m                       give character min and max bounds\n"
            "    -C                       force columns\n"
            "    -1                       force single column\n"
            "    -u                       keep output unsorted\n"
            "    -o                       use OpenFont/QueryFont instead of ListFonts\n"
            "    -w width                 maximum width for multiple columns\n"
            "    -n columns               number of columns if multi column\n"
            "    -display displayname     X server to contact\n"
            "    -d displayname           (alias for -display displayname)\n"
            "    -v                       print program version\n"
            "\n");
    Close_Display();
    exit(exitstatus);
}

int
main(int argc, char **argv)
{
    int argcnt = 0, i;

    INIT_NAME;

    /* Handle args that don't require opening a display */
    for (int a = 1; a < argc; a++) {
        const char *argn = argv[a];
        /* accept single or double dash for -help & -version */
        if (argn[0] == '-' && argn[1] == '-') {
            argn++;
        }
        if (strcmp(argn, "-help") == 0) {
            usage(NULL, EXIT_SUCCESS);
        }
        if ((strcmp(argn, "-version") == 0) ||
            (strcmp(argv[a], "-v") == 0)) {
            puts(PACKAGE_STRING);
            exit(EXIT_SUCCESS);
        }
    }

    /* Handle command line arguments, open display */
    Setup_Display_And_Screen(&argc, argv);

    for (argv++, argc--; argc; argv++, argc--) {
        if (argv[0][0] == '-') {
            if (argcnt > 0)
                usage("options may not be specified after font names",
                      EXIT_FAILURE);
            for (i = 1; argv[0][i]; i++) {
                switch (argv[0][i]) {
                case 'l':
                    long_list++;
                    break;
                case 'm':
                    min_max++;
                    break;
                case 'C':
                    columns = 0;
                    break;
                case '1':
                    columns = 1;
                    break;
                case 'f':      /* "-fn" */
                    if (argv[0][i + 1] != 'n') {
                        fprintf(stderr, "%s: unrecognized argument %s\n\n",
                                program_name, argv[0]);
                        usage(NULL, EXIT_FAILURE);
                    }
                    if (--argc <= 0)
                        usage("-fn requires an argument", EXIT_FAILURE);
                    argcnt++;
                    argv++;
                    get_list(argv[0]);
                    goto next;
                case 'w':
                    if (--argc <= 0)
                        usage("-w requires an argument", EXIT_FAILURE);
                    argv++;
                    max_output_line_width = atoi(argv[0]);
                    goto next;
                case 'n':
                    if (--argc <= 0)
                        usage("-n requires an argument", EXIT_FAILURE);
                    argv++;
                    columns = atoi(argv[0]);
                    goto next;
                case 'o':
                    open_instead_of_list = True;
                    break;
                case 'u':
                    sort_output = False;
                    break;
                case 'v':
                    puts(PACKAGE_STRING);
                    exit(0);
                default:
                    fprintf(stderr, "%s: unrecognized argument -%c\n\n",
                            program_name, argv[0][i]);
                    usage(NULL, EXIT_FAILURE);
                    break;
                }
            }
            if (i == 1) {
                fprintf(stderr, "%s: unrecognized argument %s\n\n",
                        program_name, argv[0]);
                usage(NULL, EXIT_FAILURE);
            }
        }
        else {
            argcnt++;
            get_list(argv[0]);
        }
 next: ;
    }

    if (argcnt == 0)
        get_list("*");

    show_fonts();

    Close_Display();
    return EXIT_SUCCESS;
}

static void
get_list(const char *pattern)
{
    XFontStruct *info;

    if (open_instead_of_list) {
        info = XLoadQueryFont(dpy, pattern);

        if (info == NULL) {
            fprintf(stderr, "%s: pattern \"%s\" unmatched\n",
                    program_name, pattern);
            return;
        }

        font_list = reallocarray(font_list, (font_cnt + 1), sizeof(FontList));
        if (font_list == NULL)
            Fatal_Error("Out of memory!");
        font_list[font_cnt].name = pattern;
        if (long_list == L_MEDIUM) {
            font_list[font_cnt].info = info;
            XUnloadFont(dpy, info->fid);
        }
        else {
            font_list[font_cnt].info = NULL;
            XFreeFont(dpy, info);
        }
        font_cnt++;
    }
    else {
        /* Get list of fonts matching pattern */
        int available = nnames + 1;
        char **fonts;

        for (;;) {
            if (long_list == L_MEDIUM)
                fonts = XListFontsWithInfo(dpy, pattern, nnames, &available,
                                           &info);
            else
                fonts = XListFonts(dpy, pattern, nnames, &available);
            if (fonts == NULL) {
                fprintf(stderr, "%s: pattern \"%s\" unmatched\n",
                        program_name, pattern);
                return;
            }
            if (available < nnames)
                break;
            if (long_list == L_MEDIUM)
                XFreeFontInfo(fonts, info, available);
            else
                XFreeFontNames(fonts);
            nnames = available * 2;
        }

        font_list = reallocarray(font_list,
                                 (font_cnt + available), sizeof(FontList));
        if (font_list == NULL)
            Fatal_Error("Out of memory!");
        for (int i = 0; i < available; i++) {
            font_list[font_cnt].name = fonts[i];
            if (long_list == L_MEDIUM)
                font_list[font_cnt].info = info + i;
            else
                font_list[font_cnt].info = NULL;

            font_cnt++;
        }
    }
}

static int
compare(const void *arg1, const void *arg2)
{
    const FontList *f1 = arg1;
    const FontList *f2 = arg2;
    const char *p1 = f1->name;
    const char *p2 = f2->name;

    return strcmp(p1, p2);
}

static void
show_fonts(void)
{
    int i;

    if (font_cnt == 0)
        return;

    /* first sort the output */
    if (sort_output)
        qsort(font_list, font_cnt, sizeof(FontList), compare);

    if (long_list > L_MEDIUM) {
        for (i = 0; i < font_cnt; i++) {
            do_query_font(dpy, font_list[i].name);
        }
        return;
    }

    if (long_list == L_MEDIUM) {
        XFontStruct *pfi;

        const char *string;

        printf("DIR  ");
        printf("MIN  ");
        printf("MAX ");
        printf("EXIST ");
        printf("DFLT ");
        printf("PROP ");
        printf("ASC ");
        printf("DESC ");
        printf("NAME");
        printf("\n");
        for (i = 0; i < font_cnt; i++) {
            pfi = font_list[i].info;
            if (!pfi) {
                fprintf(stderr, "%s:  no font information for font \"%s\".\n",
                        program_name,
                        font_list[i].name ? font_list[i].name : "");
                continue;
            }
            switch (pfi->direction) {
            case FontLeftToRight:
                string = "-->";
                break;
            case FontRightToLeft:
                string = "<--";
                break;
            default:
                string = "???";
                break;
            }
            printf("%-4s", string);
            if (pfi->min_byte1 == 0 && pfi->max_byte1 == 0) {
                printf(" %3d ", pfi->min_char_or_byte2);
                printf(" %3d ", pfi->max_char_or_byte2);
            }
            else {
                printf("*%3d ", pfi->min_byte1);
                printf("*%3d ", pfi->max_byte1);
            }
            printf("%5s ", pfi->all_chars_exist ? "all" : "some");
            printf("%4d ", pfi->default_char);
            printf("%4d ", pfi->n_properties);
            printf("%3d ", pfi->ascent);
            printf("%4d ", pfi->descent);
            printf("%s\n", font_list[i].name);
            if (min_max) {
                char min[BUFSIZ], max[BUFSIZ];

                char *pmax = max, *pmin = min;

                strcpy(pmin, "     min(l,r,w,a,d) = (");
                strcpy(pmax, "     max(l,r,w,a,d) = (");
                pmin += strlen(pmin);
                pmax += strlen(pmax);

                copy_number(&pmin, &pmax,
                            pfi->min_bounds.lbearing,
                            pfi->max_bounds.lbearing);
                *pmin++ = *pmax++ = ',';
                copy_number(&pmin, &pmax,
                            pfi->min_bounds.rbearing,
                            pfi->max_bounds.rbearing);
                *pmin++ = *pmax++ = ',';
                copy_number(&pmin, &pmax,
                            pfi->min_bounds.width,
                            pfi->max_bounds.width);
                *pmin++ = *pmax++ = ',';
                copy_number(&pmin, &pmax,
                            pfi->min_bounds.ascent,
                            pfi->max_bounds.ascent);
                *pmin++ = *pmax++ = ',';
                copy_number(&pmin, &pmax,
                            pfi->min_bounds.descent,
                            pfi->max_bounds.descent);
                *pmin++ = *pmax++ = ')';
                *pmin = *pmax = '\0';
                printf("%s\n", min);
                printf("%s\n", max);
            }
        }
        return;
    }

    if ((columns == 0 && isatty(1)) || columns > 1) {
        int width, max_width = 0, lines_per_column, j, index;

        for (i = 0; i < font_cnt; i++) {
            width = strlen(font_list[i].name);
            if (width > max_width)
                max_width = width;
        }
        if (max_width == 0)
            Fatal_Error("all %d fontnames listed are zero length", font_cnt);

        if (columns == 0) {
            if ((max_width * 2) + output_line_padding >
                max_output_line_width) {
                columns = 1;
            }
            else {
                max_width += output_line_padding;
                columns = ((max_output_line_width +
                            output_line_padding) / max_width);
            }
        }
        else {
            max_width += output_line_padding;
        }
        if (columns <= 1)
            goto single_column;

        if (font_cnt < columns)
            columns = font_cnt;
        lines_per_column = (font_cnt + columns - 1) / columns;

        for (i = 0; i < lines_per_column; i++) {
            for (j = 0; j < columns; j++) {
                index = j * lines_per_column + i;
                if (index >= font_cnt)
                    break;
                if (j + 1 == columns)
                    printf("%s", font_list[index].name);
                else
                    printf("%-*s", max_width, font_list[index].name);
            }
            printf("\n");
        }
        return;
    }

 single_column:
    for (i = 0; i < font_cnt; i++)
        printf("%s\n", font_list[i].name);
}

static void
copy_number(char **pp1, char **pp2, int n1, int n2)
{
    char *p1 = *pp1;
    char *p2 = *pp2;
    int w;

    sprintf(p1, "%d", n1);
    sprintf(p2, "%d", n2);
    w = MAX(strlen(p1), strlen(p2));
    sprintf(p1, "%*d", w, n1);
    sprintf(p2, "%*d", w, n2);
    p1 += strlen(p1);
    p2 += strlen(p2);
    *pp1 = p1;
    *pp2 = p2;
}

/* ARGSUSED */
static int
IgnoreError(_X_UNUSED Display *disp, _X_UNUSED XErrorEvent *event)
{
    return 0;
}

static const char *bounds_metrics_title =
    "width left  right  asc  desc   attr   keysym\n";

#define PrintBounds(_what,_ptr) \
{   register XCharStruct *p = (_ptr); \
    printf ("\t%3s\t\t%4d  %4d  %4d  %4d  %4d  0x%04x\n", \
          (_what), p->width, p->lbearing, \
          p->rbearing, p->ascent, p->descent, p->attributes); }

static const char *stringValued[] = {   /* values are atoms */
    /* font name components (see section 3.2 of the XLFD) */
    "FOUNDRY",
    "FAMILY_NAME",
    "WEIGHT_NAME",
    "SLANT",
    "SETWIDTH_NAME",
    "ADD_STYLE_NAME",
    "SPACING",
    "CHARSET_REGISTRY",
    "CHARSET_ENCODING",

    /* other standard X font properties (see section 3.2 of the XLFD) */
    "FONT",
    "FACE_NAME",
    "FULL_NAME",                /* deprecated */
    "COPYRIGHT",
    "NOTICE",
    "FONT_TYPE",
    "FONT_VERSION",
    "RASTERIZER_NAME",
    "RASTERIZER_VERSION",

    /* other registered font properties (see the X.org Registry, sec. 15) */
    "_ADOBE_POSTSCRIPT_FONTNAME",

    /* unregistered font properties */
    "CHARSET_COLLECTIONS",
    "CLASSIFICATION",
    "DEVICE_FONT_NAME",
    "FONTNAME_REGISTRY",
    "MONOSPACED",
    "QUALITY",
    "RELATIVE_SET",
    "STYLE",
    NULL
};

static void
PrintProperty(XFontProp * prop)
{
    char *atom, *value;
    char nosuch[40];
    int i;
    XErrorHandler oldhandler = XSetErrorHandler(IgnoreError);

    atom = XGetAtomName(dpy, prop->name);
    if (!atom) {
        snprintf(nosuch, sizeof(nosuch), "No such atom = %ld", prop->name);
        atom = nosuch;
    }
    printf("      %s", atom);

    /* Pad out to a column width of 22, but ensure there is always at
       least one space between property name & value. */
    for (i = strlen(atom); i < 21; i++)
        putchar(' ');
    putchar(' ');

    for (i = 0;; i++) {
        if (stringValued[i] == NULL) {
            printf("%ld\n", prop->card32);
            break;
        }
        if (strcmp(stringValued[i], atom) == 0) {
            value = XGetAtomName(dpy, prop->card32);
            if (value == NULL)
                printf("%ld (expected string value)\n", prop->card32);
            else {
                printf("%s\n", value);
                XFree(value);
            }
            break;
        }
    }
    if (atom != nosuch)
        XFree(atom);
    XSetErrorHandler(oldhandler);
}

static void
ComputeFontType(XFontStruct *fs)
{
    Bool char_cell = True;
    const char *reason = NULL;
    XCharStruct *cs;
    Atom awatom = XInternAtom(dpy, "AVERAGE_WIDTH", False);

    printf("  font type:\t\t");
    if (fs->min_bounds.width != fs->max_bounds.width) {
        printf("Proportional (min and max widths not equal)\n");
        return;
    }

    if (awatom) {
        for (int i = 0; i < fs->n_properties; i++) {
            if (fs->properties[i].name == awatom &&
                (fs->max_bounds.width * 10) != fs->properties[i].card32) {
                char_cell = False;
                reason = "font width not equal to AVERAGE_WIDTH";
                break;
            }
        }
    }

    if (fs->per_char) {
        unsigned int i;
        for (i = fs->min_char_or_byte2, cs = fs->per_char;
             i <= fs->max_char_or_byte2; i++, cs++) {
            if (cs->width == 0)
                continue;
            if (cs->width != fs->max_bounds.width) {
                /* this shouldn't happen since we checked above */
                printf("Proportional (characters not all the same width)\n");
                return;
            }
            if (char_cell) {
                if (cs->width < 0) {
                    if (!(cs->width <= cs->lbearing &&
                          cs->lbearing <= cs->rbearing &&
                          cs->rbearing <= 0)) {
                        char_cell = False;
                        reason = "ink outside bounding box";
                    }
                }
                else {
                    if (!(0 <= cs->lbearing &&
                          cs->lbearing <= cs->rbearing &&
                          cs->rbearing <= cs->width)) {
                        char_cell = False;
                        reason = "ink outside bounding box";
                    }
                }
                if (!(cs->ascent <= fs->ascent &&
                      cs->descent <= fs->descent)) {
                    char_cell = False;
                    reason = "characters not all same ascent or descent";
                }
            }
        }
    }

    printf("%s", char_cell ? "Character Cell" : "Monospaced");
    if (reason)
        printf(" (%s)", reason);
    printf("\n");

    return;
}

static void
print_character_metrics(register XFontStruct *info)
{
    register XCharStruct *pc = info->per_char;
    unsigned int i, j;
    unsigned n, saven;

    printf("  character metrics:\n");
    saven = ((info->min_byte1 << 8) | info->min_char_or_byte2);
    for (j = info->min_byte1; j <= info->max_byte1; j++) {
        n = saven;
        for (i = info->min_char_or_byte2; i <= info->max_char_or_byte2; i++) {
            char *s = XKeysymToString((KeySym) n);

            printf("\t0x%02x%02x (%u)\t%4d  %4d  %4d  %4d  %4d  0x%04x  %s\n",
                   j, i, n, pc->width, pc->lbearing,
                   pc->rbearing, pc->ascent, pc->descent, pc->attributes,
                   s ? s : ".");
            pc++;
            n++;
        }
        saven += 256;
    }
}

static void
do_query_font(Display *display, const char *name)
{
    register int i;
    register XFontStruct *info = XLoadQueryFont(display, name);

    if (!info) {
        fprintf(stderr, "%s:  unable to get info about font \"%s\"\n",
                program_name, name);
        return;
    }
    printf("name:  %s\n", name ? name : "(nil)");
    printf("  direction:\t\t%s\n", ((info->direction == FontLeftToRight)
                                    ? "left to right" : "right to left"));
    printf("  indexing:\t\t%s\n",
           ((info->min_byte1 == 0 && info->max_byte1 == 0)
            ? "linear" : "matrix"));
    printf("  rows:\t\t\t0x%02x thru 0x%02x (%d thru %d)\n",
           info->min_byte1, info->max_byte1,
           info->min_byte1, info->max_byte1);
    printf("  columns:\t\t0x%02x thru 0x%02x (%d thru %d)\n",
           info->min_char_or_byte2, info->max_char_or_byte2,
           info->min_char_or_byte2, info->max_char_or_byte2);
    printf("  all chars exist:\t%s\n",
           (info->all_chars_exist) ? "yes" : "no");
    printf("  default char:\t\t0x%04x (%d)\n",
           info->default_char, info->default_char);
    printf("  ascent:\t\t%d\n", info->ascent);
    printf("  descent:\t\t%d\n", info->descent);
    ComputeFontType(info);
    printf("  bounds:\t\t%s", bounds_metrics_title);
    PrintBounds("min", &info->min_bounds);
    PrintBounds("max", &info->max_bounds);
    if (info->per_char && long_list >= L_VERYLONG)
        print_character_metrics(info);
    printf("  properties:\t\t%d\n", info->n_properties);
    for (i = 0; i < info->n_properties; i++)
        PrintProperty(&info->properties[i]);
    printf("\n");

    XFreeFontInfo(NULL, info, 1);
}
